File: cpp/tests/test_vec.py
import pytest
from vec import Vec3i, Vec3f
import numpy as np
import numpy.testing as npt
    
class TestVec3:
    
    @pytest.fixture(name='zero')
    def test_vec3_default(self):
        return Vec3i()
    
    @pytest.fixture(name='one_i')
    def test_vec3_one_i(self):
        return Vec3i(1,1,1)
        
    @pytest.fixture(name='one_f')
    def test_vec3f_one(self):
        return Vec3f(1., 1., 1.)
    
    def test_vec3_repr(self, one_i):
        assert repr(one_i) == '<Vec3(1, 1, 1)>'
        
    def test_vec3_setter_getter(self, one_i):
        one_i.x = 2
        assert one_i.x == 2
    
    def test_vec3f(self, one_f):
        one_f.x = 1.1
        npt.assert_allclose(one_f.x, 1.1)
        
    def test_vec3_numpy(self, one_i):
        npt.assert_equal(np.array(one_i), np.array([1,1,1]))
        
    def test_vec3f_numpy(self, one_f):
        npt.assert_allclose(np.array(one_f), np.array([1.,1.,1.]))

File: cpp/tests/test_vec.cpp
#include "vec.h"
#include <gtest/gtest.h>

TEST(testVec3, default_construct)
{
    Vec3<int> v;
    EXPECT_EQ(0, v.x);
    EXPECT_EQ(0, v.y);
    EXPECT_EQ(0, v.z);
}

TEST(testVec3, index)
{
    Vec3<int> v = {1,2,3};
    EXPECT_EQ(1, v[0]);
    EXPECT_EQ(2, v[1]);
    EXPECT_EQ(3, v[2]);
}